File: ios/Sources/KlinikAPI/NotificationsAPI.swift
import Foundation
import KlinikCore

public enum NotificationChannel: String, Decodable, Encodable, Sendable, CaseIterable {
    case push = "PUSH"
    case sms = "SMS"
    case email = "EMAIL"
    case whatsapp = "WHATSAPP"
    case inApp = "IN_APP"

    public var localizedName: String { L10n.string("notification.channel.\(rawValue)") }
}

public enum NotificationDeliveryStatus: String, Decodable, Sendable, Equatable {
    case pending = "PENDING"
    case sent = "SENT"
    case delivered = "DELIVERED"
    case read = "READ"
    case failed = "FAILED"

    public var localizedName: String { L10n.string("notification.status.\(rawValue)") }
}

/// The notification types a person can turn on or off, mirroring the server.
public enum NotificationKind: String, Decodable, Encodable, Sendable, CaseIterable {
    case labReady = "lab.ready"
    case labCritical = "lab.critical"
    case newMessage = "message.new"
    case medicationDue = "medication.due"
    case appointmentReminder = "appointment.reminder"
    case documentMissing = "document.missing"
    case complicationAnswered = "complication.answered"

    public var localizedName: String { L10n.string("notification.type.\(rawValue)") }
}

public struct NotificationPreference: Decodable, Sendable, Equatable, Identifiable {
    public let type: String
    public let channel: NotificationChannel
    public let enabled: Bool
    public let quietHoursStart: String?
    public let quietHoursEnd: String?
    public let timezone: String

    public var id: String { "\(type)|\(channel.rawValue)" }
}

public struct DeliveredNotification: Decodable, Sendable, Equatable, Identifiable {
    public let id: String
    public let type: String
    public let title: String
    public let body: String
    public let channel: NotificationChannel
    public let status: NotificationDeliveryStatus
    /// Why this attempt did not arrive.
    public let failureReason: String?
    /// The attempt this one is standing in for.
    public let fallbackForId: String?
    public let sentAt: Date?
    public let readAt: Date?
    public let createdAt: Date

    public var isFallback: Bool { fallbackForId != nil }
}

public struct NotificationsAPI: Sendable {
    private let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    /// Registers this device. Called after the system grants permission, and
    /// again whenever the platform hands out a new token.
    public func registerToken(_ token: String, platform: String, deviceId: String?) async throws {
        try await client.send(
            Endpoint(
                method: .post,
                path: "me/notifications/tokens",
                body: try JSONEncoder.klinik.encode(
                    TokenBody(token: token, platform: platform, deviceId: deviceId)
                )
            )
        )
    }

    /// Called on sign-out, so a device stops receiving what it may no longer see.
    public func revokeToken(_ token: String) async throws {
        try await client.send(
            Endpoint(method: .delete, path: "me/notifications/tokens", query: ["token": token])
        )
    }

    public func preferences() async throws -> [NotificationPreference] {
        try await client.send(
            Endpoint(method: .get, path: "me/notifications/preferences"),
            as: [NotificationPreference].self
        )
    }

    @discardableResult
    public func setPreference(
        type: NotificationKind,
        channel: NotificationChannel,
        enabled: Bool,
        quietHoursStart: String? = nil,
        quietHoursEnd: String? = nil
    ) async throws -> NotificationPreference {
        try await client.send(
            Endpoint(
                method: .put,
                path: "me/notifications/preferences",
                body: try JSONEncoder.klinik.encode(
                    PreferenceBody(
                        type: type.rawValue,
                        channel: channel,
                        enabled: enabled,
                        quietHoursStart: quietHoursStart,
                        quietHoursEnd: quietHoursEnd
                    )
                )
            ),
            as: NotificationPreference.self
        )
    }

    public func history() async throws -> [DeliveredNotification] {
        try await client.send(
            Endpoint(method: .get, path: "me/notifications"),
            as: [DeliveredNotification].self
        )
    }

    @discardableResult
    public func markRead() async throws -> Int {
        let result = try await client.send(
            Endpoint(method: .post, path: "me/notifications/read"),
            as: MarkedRead.self
        )

        return result.marked
    }

    private struct TokenBody: Encodable {
        let token: String
        let platform: String
        let deviceId: String?
    }

    private struct PreferenceBody: Encodable {
        let type: String
        let channel: NotificationChannel
        let enabled: Bool
        let quietHoursStart: String?
        let quietHoursEnd: String?
    }

    private struct MarkedRead: Decodable, Sendable {
        let marked: Int
    }
}

File: android/core/network/src/main/kotlin/xyz/klinik/network/NotificationsApi.kt
package xyz.klinik.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
enum class NotificationChannel {
    PUSH,
    SMS,
    EMAIL,
    WHATSAPP,
    IN_APP,
    ;

    val stringKey: String get() = "notification_channel_${name.lowercase()}"
}

@Serializable
enum class NotificationDeliveryStatus {
    PENDING,
    SENT,
    DELIVERED,
    READ,
    FAILED,
    ;

    val stringKey: String get() = "notification_status_${name.lowercase()}"
}

/** The notification types a person can turn on or off, mirroring the server. */
enum class NotificationKind(val wire: String) {
    LAB_READY("lab.ready"),
    LAB_CRITICAL("lab.critical"),
    NEW_MESSAGE("message.new"),
    MEDICATION_DUE("medication.due"),
    APPOINTMENT_REMINDER("appointment.reminder"),
    DOCUMENT_MISSING("document.missing"),
    COMPLICATION_ANSWERED("complication.answered"),
    ;

    val stringKey: String get() = "notification_type_${wire.replace('.', '_')}"
}

@Serializable
data class NotificationPreference(
    val type: String,
    val channel: NotificationChannel,
    val enabled: Boolean,
    val quietHoursStart: String? = null,
    val quietHoursEnd: String? = null,
    val timezone: String = "Europe/Istanbul",
)

@Serializable
data class DeliveredNotification(
    val id: String,
    val type: String,
    val title: String,
    val body: String,
    val channel: NotificationChannel,
    val status: NotificationDeliveryStatus,
    /** Why this attempt did not arrive. */
    val failureReason: String? = null,
    /** The attempt this one is standing in for. */
    val fallbackForId: String? = null,
    val sentAt: String? = null,
    val readAt: String? = null,
    val createdAt: String,
) {
    val isFallback: Boolean get() = fallbackForId != null
}

@Serializable
private data class TokenBody(
    val token: String,
    val platform: String,
    val deviceId: String? = null,
)

@Serializable
private data class PreferenceBody(
    val type: String,
    val channel: NotificationChannel,
    val enabled: Boolean,
    val quietHoursStart: String? = null,
    val quietHoursEnd: String? = null,
)

@Serializable
private data class MarkedNotificationsRead(val marked: Int)

class NotificationsApi(
    private val client: ApiClient,
    private val json: Json = ApiClient.defaultJson,
) {
    /**
     * Registers this device. Called after the system grants permission, and
     * again whenever the platform hands out a new token.
     */
    suspend fun registerToken(token: String, platform: String = "android", deviceId: String? = null) {
        client.send(
            Endpoint(
                HttpMethod.POST,
                "me/notifications/tokens",
                body = json.encodeToString(
                    TokenBody.serializer(),
                    TokenBody(token, platform, deviceId),
                ),
            ),
        )
    }

    /** Called on sign-out, so a device stops receiving what it may no longer see. */
    suspend fun revokeToken(token: String) {
        client.send(
            Endpoint(
                HttpMethod.DELETE,
                "me/notifications/tokens",
                query = mapOf("token" to token),
            ),
        )
    }

    suspend fun preferences(): List<NotificationPreference> =
        decode(client.send(Endpoint(HttpMethod.GET, "me/notifications/preferences")))

    suspend fun setPreference(
        type: NotificationKind,
        channel: NotificationChannel,
        enabled: Boolean,
        quietHoursStart: String? = null,
        quietHoursEnd: String? = null,
    ): NotificationPreference =
        decode(
            client.send(
                Endpoint(
                    HttpMethod.PUT,
                    "me/notifications/preferences",
                    body = json.encodeToString(
                        PreferenceBody.serializer(),
                        PreferenceBody(type.wire, channel, enabled, quietHoursStart, quietHoursEnd),
                    ),
                ),
            ),
        )

    suspend fun history(): List<DeliveredNotification> =
        decode(client.send(Endpoint(HttpMethod.GET, "me/notifications")))

    suspend fun markRead(): Int =
        decode<MarkedNotificationsRead>(
            client.send(Endpoint(HttpMethod.POST, "me/notifications/read")),
        ).marked

    private inline fun <reified T> decode(body: String): T =
        runCatching { json.decodeFromString<T>(body) }
            .getOrElse { throw ApiError.Decoding(it.message ?: "unreadable response") }
}
